metabase/metabase
# Migrate notebook aggregations to MLv2 — Metrics (2)

Part of #31001

The second step in the notebook editor's `AggregateStep` migration. Adds support for metrics (distinct entities that can be added in "Admin > Data Model").

We're about to rework Metrics and make them model-based. It didn't make a lot of sense to port legacy metrics to MLv2, so we're still using MLv1 here. I'm open for feedback on how to approach this (look at `legacyQuery` and `onChangeLegacy` props in `AggregationPicker`)

> **Warning**
> This PR targets a feature branch and doesn't migrate `AggregateStep` features 1:1
> This branch is expected to have failing tests because parts of functionality are simply missing
> You can monitor the migration health with the last branch in the chain — #31530

### To Verify

1. Sign in as Metabase admin
2. Open `/admin/datamodel/metrics` and add a few metrics
3. For different kinds of questions (structured/native, based on raw data/models/questions, flat/multi-stage):
   1. Open `/question/:id/notebook`
   2. Click "Pick the metric you want to see"
   3. Ensure you can see, apply, change, and remove metrics you've created at step 2
   4. Overall ensure the behavior matches `master`

### Demo

https://github.com/metabase/metabase/assets/17258145/dcb38d9e-52fa-4689-aaec-94cf0246cc42
